File: MissingNumber.py
"""
268.缺失数字
给定一个包含 0, 1, 2, ..., n 中 n 个数的序列，找出 0 .. n 中没有出现在序列中的那个数。
示例 1:
输入: [3,0,1]
输出: 2
示例 2:
输入: [9,6,4,2,3,5,7,0,1]
输出: 8
说明:
你的算法应具有线性时间复杂度。你能否仅使用额外常数空间来实现?
来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/missing-number
著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
"""
class Solution:
    #5.02%,25.31%
    def missingNumber(self, nums) -> int:
        for i in range(len(nums)+1):
            if i not in nums:
                return i

    #95.33%, 22.94%
    #利用数学法，但有可能造成溢出，最好一边加一边减
    def missingNumber2(self, nums) -> int:
        sum = 0
        isHaveZero = False
        for num in nums:
            if num == 0:
                isHaveZero = True
            sum += num
        if not isHaveZero:
            return 0
        maxNum = max(nums)
        temp = (maxNum * (maxNum + 1) >> 1)
        if temp == sum:
            return maxNum + 1
        return temp - sum
    #改善上面的数学法
    #96.94%，25.31%
    def missingNumber3(self, nums) -> int:
        sum = 0
        for i, num in enumerate(nums):
            sum += i+1
            sum -= num
        return sum

    #位运算法
    def missingNumber4(self, nums) -> int:
        miss = len(nums)
        for i, num in enumerate(nums):
            miss ^= i ^ num
        return miss

s = Solution()
print(s.missingNumber4([1,0]))
print(s.missingNumber4([1]))
print(s.missingNumber4([0]))
print(s.missingNumber4([3,0,1]))
print(s.missingNumber4([9,6,4,2,3,5,7,0,1]))

File: src/MissingNumber.java
import java.util.Arrays;

/**面试题17.04.消失的数字
 * 数组nums包含从0到n的所有整数，但其中缺了一个。请编写代码找出那个缺失的整数。你有办法在O(n)时间内完成吗？*/
public class MissingNumber {
    //11.92
    public int missingNumber(int[] nums) {
        Arrays.sort(nums);
        for (int i = 0; i < nums.length; i++) {
            if (i != nums[i]) {
                return i;
            }
        }
        return nums.length;
    }

    //62.54
    public int missingNumber2(int[] nums) {
        int[] dic = new int[nums.length+1];
        for (int num : nums) {
            dic[num]++;
        }
        for (int i = 0; i < dic.length; i++) {
            if (dic[i] == 0) {
                return i;
            }
        }
        return nums.length;
    }

    //100
    public int missingNumber3(int[] nums) {
        int r_sum = 0;
        int sum = 0;
        for (int i = 0; i < nums.length; i++){
            r_sum += nums[i];
            sum += i;
        }
        sum += nums.length;
        return sum - r_sum;
    }

    //优秀解答
    //对同一个值异或两次，那么结果等于它本身，所以我们对res从0-nums.length进行异或，
    // 同时对nums数组中的值进行异或，出现重复的会消失，所以最后res的值是只出现一次的数字，也就是nums数组中缺失的那个数字。
    //作者：LSZ
    //链接：https://leetcode-cn.com/problems/missing-number-lcci/solution/onshi-jian-fu-za-du-o1kong-jian-fu-za-du-shi-xian-/
    //来源：力扣（LeetCode）
    //著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。
    public int missingNumber4(int[] nums) {
        int res = 0;
        for (int i = 0; i < nums.length; ++i) {
            res ^= i;
            res ^= nums[i];
        }
        res ^= nums.length;

        return res;
    }

}
